### PR TITLE
Imperative workflow declaration

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Introduce a new `ImperativeWorkflow` class that can be used to build workflows up iteratively, to be used for programmatic workflow creation.

Cookbook: https://github.com/flyteorg/flytesnacks/pull/138/files

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
* Rename the `Workflow` class to `PythonFunctionWorkflow`.
* Extract out the common parts of the old `Workflow` class logic into a `WorkflowBase` class.
* Add a new `ImperativeWorkflow` class (but it's imported `as Workflow` in `flytekit/__init__.py` since users shouldn't really ever have to concern themselves with the `PythonFunctionWorkflow` class.
* `python_interface` is the property that pulls in our new-api python `Interface` class.  Changed references to `wf._native_interface` to use this instead.
* Separating out the `_local_execute` from the `execute` function in the workflow classes lets us also patch the imperative workflow correctly.

Other bugs:
* `with_overrides` in the node object was not returning self
* Type engine was not erroring when trying to convert a None.

## Tracking Issue
NA

### CODE REVIEW

File: flytekit/core/workflow.py [hunk 1]
```diff
@@ -135,98 +163,114 @@ def name(self) -> str:
...
+            return create_and_link_node(ctx, entity=self, interface=self.python_interface, **input_kwargs)
 
-            # This needs a longer comment.
-            for n in comp_ctx.compilation_state.nodes:
-                if isinstance(n.flyte_entity, PythonAutoContainerTask) and n.flyte_entity.task_resolver == self:
-                    logger.debug(f"WF {self.name} saving task {n.flyte_entity.name}")
-                    self.add(n.flyte_entity)
+        # Next is executing as a subworkflow
+        elif (
+            ctx.execution_state is not None and ctx.execution_state.mode == ExecutionState.Mode.LOCAL_WORKFLOW_EXECUTION
```
Comment (katrogan) [inline flytekit/core/workflow.py:212]: is this true? it's executing as a subworkflow for LOCAL_WORKFLOW_EXECUTION

File: flytekit/core/workflow.py [hunk 2]
```diff
@@ -135,98 +163,114 @@ def name(self) -> str:
...
+            if expected_outputs == 0:
+                if result is None or isinstance(result, VoidPromise):
+                    return None
+                else:
+                    raise Exception(f"Workflow local execution expected 0 outputs but something received {result}")
 
+            if (expected_outputs > 1 and len(result) == expected_outputs) or (
+                expected_outputs == 1 and result is not None
+            ):
+                if isinstance(result, Promise):
```
Comment (katrogan) [inline flytekit/core/workflow.py:250]: perhaps a stupid q, but can a result be anything but a promise when expected_outputs == 1?

File: flytekit/core/workflow.py [hunk 3]
```diff
@@ -277,87 +320,322 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
...
+        return self._compilation_state
+
+    @property
+    def nodes(self) -> List[Node]:
+        return self._compilation_state.nodes
+
+    @property
+    def inputs(self) -> Dict[str, Promise]:
+        """
+        This holds the input promises to the workflow. The nodes in these Promise objects should always be from
```
Comment (katrogan) [inline flytekit/core/workflow.py:360]: nit: The **nodes** in these Promise objects should always be from (?)

File: flytekit/core/workflow.py [hunk 4]
```diff
@@ -277,87 +320,322 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
...
+            return n
 
-            with ctx.new_execution_context(mode=ExecutionState.Mode.LOCAL_WORKFLOW_EXECUTION) as ctx:
-                result = self._local_execute(ctx, **kwargs)
+    def add_workflow_input(self, input_name: str, python_type: Type) -> Interface:
+        """
+        Adds an input to the workflow.
+        """
+        if input_name in self._inputs:
+            raise FlyteValueException("already in")
```
Comment (katrogan) [inline flytekit/core/workflow.py:481]: nit: more detailed exception?

File: flytekit/core/workflow.py [hunk 5]
```diff
@@ -277,87 +320,322 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
...
-            expected_outputs = len(self._native_interface.outputs)
-            if expected_outputs == 0:
-                if result is None or isinstance(result, VoidPromise):
-                    return None
-                else:
-                    raise Exception(f"Workflow local execution expected 0 outputs but something received {result}")
+        if python_type is None:
+            if type(p) == list or type(p) == dict:
+                raise FlyteValidationException(
+                    f"If not a straight up promise, wf output {output_name} needs to be given an explicit type."
```
Comment (katrogan) [inline flytekit/core/workflow.py:500]:
```suggestion
                    f"If not a straight up promise, wf output {output_name} needs to be given a scalar type."
```

Comment (katrogan) [reply flytekit/core/workflow.py:500]: if I'm understanding correctly

File: flytekit/core/workflow.py [hunk 6]
```diff
@@ -277,87 +320,322 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
...
+
+        return True
+
+
+class PythonFunctionWorkflow(WorkflowBase, ClassStorageTaskResolver):
+    """
+    When you assign a name to a node.
+
+    * Any upstream node that is not assigned, recursively assign
+    * When you get the call to the constructor, keep in mind there may be duplicate nodes, because they all should
```
Comment (katrogan) [inline flytekit/core/workflow.py:541]: what does this mean?

**Review by katrogan — COMMENTED** [review]
sorry a few questions

File: flytekit/core/workflow.py (hunk 6 above)
Comment (wild-endeavor) [reply flytekit/core/workflow.py:536]: It means this is a months old bad comment that I wrote myself and needs to be deleted.  It's only showing up now because I made a base class.  Will remove and open a separate PR to add a proper description.

File: flytekit/core/workflow.py (hunk 5 above)
Comment (wild-endeavor) [reply flytekit/core/workflow.py:495]: Actually the opposite, will clarify.

File: flytekit/core/workflow.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/core/workflow.py:212]: yeah. will clarify.

File: flytekit/core/workflow.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/core/workflow.py:250]: Yeah, a workflow can be defined as:

```
def wf() -> int:
  t1()
  return 5
```

File: flytekit/core/workflow.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/core/workflow.py:360]: huh?

Comment (wild-endeavor) [reply flytekit/core/workflow.py:360]: I'll reword a bit, but not sure why "nodes" was in bold.

Comment (katrogan) [reply flytekit/core/workflow.py]: oh sorry i was confused by this comment

File: flytekit/core/type_engine.py [hunk 7]
```diff
@@ -262,6 +262,8 @@ def to_literal_type(cls, python_type: Type) -> LiteralType:
 
     @classmethod
     def to_literal(cls, ctx: FlyteContext, python_val: typing.Any, python_type: Type, expected: LiteralType) -> Literal:
+        if python_val is None:
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:265]: @kumare3 this is okay right?  I think it was just returning a Primitive earlier with nothing set.

Comment (kumare3) [reply flytekit/core/type_engine.py:265]: I think this is right. Just that we may want a none handler?
So if python_val is None and python_type is None, then it is ok to return Void?

File: tests/flytekit/unit/core/test_imperative_errors.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    image_config=ImageConfig(default_image=default_img, images=[default_img]),
+)
+
+
+@task
+def t1(a: str) -> str:
+    return a + " world"
+
+
+wb = ImperativeWorkflow(name="my.workflow")
```
Comment (kumare3) [inline tests/flytekit/unit/core/test_imperative_errors.py:26]: so we will not support branches in the imperative world?

File: flytekit/core/workflow.py [hunk 9]
```diff
@@ -91,41 +91,69 @@ def construct_input_promises(inputs: List[str]):
     }
 
 
-class Workflow(ClassStorageTaskResolver):
-    """
-    When you assign a name to a node.
+# given a binding to a binding collection, want to return a literal collection.
```
Comment (kumare3) [inline flytekit/core/workflow.py:94]: this doc does not make sense?

File: flytekit/core/workflow.py [hunk 10]
```diff
@@ -277,87 +339,319 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
...
+        all_nodes = []
+        prefix = f"{ctx.compilation_state.prefix}-{self.short_name}-" if ctx.compilation_state is not None else None
+        with ctx.new_compilation_context(prefix=prefix, task_resolver=self) as comp_ctx:
+            # Construct the default input promise bindings, but then override with the provided inputs, if any
+            input_kwargs = construct_input_promises([k for k in self.interface.inputs.keys()])
+            input_kwargs.update(kwargs)
+            workflow_outputs = self._workflow_function(**input_kwargs)
+            all_nodes.extend(comp_ctx.compilation_state.nodes)
+
+            # This needs a longer comment.
```
Comment (katrogan) [inline flytekit/core/workflow.py:605]: is this a todo or meant to be part of this PR?

**Review by katrogan — DISMISSED** [review]

File: flytekit/core/testing.py [hunk 11]
```diff
@@ -46,17 +46,23 @@ def _log(*args, **kwargs):
...
         not isinstance(target, PythonTask)
-        and not isinstance(target, Workflow)
+        and not isinstance(target, WorkflowBase)
         and not isinstance(target, ReferenceEntity)
     ):
         raise Exception("Can only use mocks on tasks/workflows declared in Python.")
 
+    logger.info(
+        "When using this patch function on Flyte entities, please be aware weird issues may arise if also"
+        "using mock.patch on internal Flyte classes like PythonFunctionWorkflow. See"
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/core/testing.py:62]: i feel like python users find every which way to sunday to break flytekit so i think i'll leave it in there for now if that's okay.  In any case having more people look at the issue is not a bad thing.

File: flytekit/core/workflow.py (hunk 10 above)
Comment (wild-endeavor) [reply flytekit/core/workflow.py:605]: not part of this PR, but i may as well add it.

File: flytekit/core/testing.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply flytekit/core/testing.py:62]: oh sure! i was just curious

**Review by katrogan — APPROVED** [review]

**Review by katrogan — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
